### DISCUSSION TITLE
Masking with a grayscale image with ImageMagick v7 and PHP

### DISCUSSION THREAD

**Discussion by Kramazeek** [ImageMagick/ImageMagick#4400] [Digital Image Processing]
I am trying to create a mask in Imagemagick (v7.0.11) that behaves like a layer mask in Photoshop where white areas are completely opaque and black areas are fully masked. Thanks for any help!
```

function grayscaleMaskTest () {

    //Create a black and white mask
    $mask = new Imagick();
    $mask->newImage(300, 300, "#ffffff");
    $mask->setImageType(IMAGICK::IMGTYPE_GRAYSCALE);
    $mask->setImageFormat('gif');

    //draw black circle
    $draw = new ImagickDraw(); 
    $draw->setStrokeWidth(1);
    $draw->setStrokeOpacity(1);
    $draw->setStrokeColor(new ImagickPixel("#000000"));
    $draw->setFillColor(new ImagickPixel("#000000"));
    $draw->circle(300/2, 300/2, 300/2, 300/2 + 100);
    $mask->drawImage($draw);
    //$mask->writeImage('C:\Users\Mike\Documents\images\mask.png'); //export mask

    //Create final image
    $final = new Imagick();
    $final->newImage(300, 300, "#FF0000");

    //Not sure if I had this right but tried writing the mask data to the alpha channel
    //$final->compositeImage($mask, IMAGICK::COMPOSITE_COPYOPACITY, 0, 0, IMAGICK::CHANNEL_ALPHA);
    $final->compositeImage($mask, IMAGICK::COMPOSITE_COPYOPACITY, 0, 0);

    $final->writeImage('C:\Users\Mike\Documents\images\final.png');
}
```

The mask looks like this :

![mask](https://user-images.githubusercontent.com/92959407/138384771-34ade6de-1022-4eca-b5c9-8d6c87d3ed8a.png)

The ideal result would look like this :

![final2](https://user-images.githubusercontent.com/92959407/138384809-70b655c3-d969-4d4e-b1dd-dd0f71b690c5.png)

**Comment by fmw42** [comment]
You can use the PHP Imagick equivalent of -negate on your mask to invert white and black and get the result you show. (see https://www.php.net/manual/en/imagick.negateimage.php) Have you tried that?

**Reply by Kramazeek** [reply]
@fmw42 - Thank you, I haven't used that and I'll file that away for the future for sure. I think the core issue I'm having is not being able to read from a grayscale image to set the transparency values on the other image think? Everything I've tried either still just copies the alpha of the mask or simply doesn't work.

**Reply by fmw42** [reply]
I still do not understand what you want to do. Please outline each step you want. Do you want to copy a grayscale image from one image and use as an alpha channel in another or in the same image?  Please note that GIF does not support 8-bit alpha, only binary alpha. So you would have to save to PNG or TIFF and not GIF if you want 8-bit alpha.

**Reply by Kramazeek** [reply]
@fwm42 - Sorry, I'm not well versed with this.

1). I need to create a grayscale image dynamically for use as a mask (copying alpha isn't an option).
2). The grayscale mask will be applied to a different image. An image with a solid fill is a good example.
3). Where the mask is white the final image will be opaque, where the mask is black the final image will be transparent.
4). Image 3 and 4 above shows the grayscale mask I think I need and the desired result.

**Reply by fmw42** [reply]
You have it already by simply using negateImage() on your current mask. Then use compositeImage() with the compose option of CopyOpacity to put that image into the alpha channel of the red image (so long as the red image and mask are the same sizes)

**Comment by snibgo** [comment]
You make a mask with:
```
$mask->newImage(250, 250, new ImagickPixel('#00000000'));
```
... so the background is transparent black. But you want opaque white.

**Reply by Kramazeek** [reply]
@snibgo - You're totally right, I know that code is messed up. I'm also copying the alpha with I don't want. If I create a black and white mask I need to be able to read those grayscale values to create transparency which I can't get to work.  I think I need to know how to create a grayscale mask with no alpha and then use those grayscale values to set transparency, right? Maybe?

**Reply by fmw42** [reply]
You use -compose copy_opacity -composit equivalent in PHP Imagick to put a grayscale image into the alpha channel of an image so that it has the grayscale values for the transparency. See https://www.php.net/manual/en/imagick.compositeimage.php

**Reply by Kramazeek** [reply]
@fmw42 - I saw that too but couldn't get it to work. Let me whip up another attempt and post the code, perhaps that will show where I'm messing up? Thanks much. I'll add more in a bit.

**Reply by Kramazeek** [reply]
@fmw42 - Like this? None of this produces the correct result? 

`function test7 () {

	$mask = new Imagick();
	$mask->newImage(400, 400, "#ffffff");
	//$mask->setImageType(IMAGICK::IMGTYPE_GRAYSCALE);
	$mask->setImageFormat('gif');

	//draw black circle
	$draw = new ImagickDraw(); 
	$draw->setStrokeWidth(1);
	$draw->setStrokeOpacity(1);
	$draw->setStrokeColor(new ImagickPixel("#000000"));
	$draw->setFillColor(new ImagickPixel("#8c8b8b"));
	$draw->circle(400/2, 400/2, 400/2, 400/2 + 150);
	$mask->drawImage($draw);
	$mask->writeImage('C:\Users\Mike\Documents\images\mask.png');	

	$final = new Imagick();
	$final->newImage(400, 400, "#FF0000");
	$final->compositeImage($mask, IMAGICK::COMPOSITE_COPYOPACITY, 0, 0, IMAGICK::CHANNEL_ALPHA);
	//$final->compositeImage($mask, IMAGICK::COMPOSITE_COPYOPACITY, 0, 0);

	$final->writeImage('C:\Users\Mike\Documents\images\final.png');
}`

**Reply by Kramazeek** [reply]
@fwm42 - I tried a few things with the items commented out to no avail. At some point I got a red circle at least so I tried changing the fill color of the mask circle to a gray to see if it effected opacity on the final image and it did not.  Different shades of gray didn't effect transparency in any way.

**Reply by fmw42** [reply]
Sorry, I do not know Imagick that well and have only tried it for a couple of simple things. So you really should be posting to an Imagick forum for answers. 

I don't think you have to set the mask type to grayscale or the format to GIF, since you save the mask as a binary (black and white only) PNG.


On the compositeImage() page down at the bottom are examples. One shows how to put an image into the alpha channel. The code is:

To copy the alpha channel from one image to another, you can do the following:

```
<?php

$img1 = new Imagick( "image1.png" );
$img2 = new Imagick( "image2.png" );

$img1->compositeImage( $img2, imagick::COMPOSITE_COPYOPACITY, 0, 0 );

header('Content-type: image/png');
echo $img1;

?>

```

I do not see anything else that is obvious to my limited knowledge of Imagick.

**Reply by Kramazeek** [reply]
@fmw42 - Thanks for taking a shot at the problem, man. I appreciate it. I'll see if anyone else chimes in or look elsewhere.

**Reply by fmw42** [reply]
Break the problem down to its components. Can you make a circle image and save it?  Once you can, then can you make a red image. Once you can, then can you put the circle into the alpha channel of the red image. See the example I posted. Do each step separately. Once they all work, then put them together in the same script.

**Reply by fmw42** [reply]
Why is your code a function? Just test as its own script as I advised above.

**Reply by fmw42** [reply]
You can try copying the grayscale image into its own alpha channel before your composite. compositeImage() copyOpacity may try to copy the alpha channel from the grayscale image. So the grayscale image may need to be put into its own alpha channel first. So see https://www.php.net/manual/en/imagick.setimagealphachannel.php and its shape option. See 

> imagick::ALPHACHANNEL_SHAPE (int)
> This constant is available if Imagick has been compiled against ImageMagick version 6.4.6 or higher.

Also see Imagemagick -alpha shape at https://legacy.imagemagick.org/Usage/masking/#alpha_shape

**Reply by Kramazeek** [reply]
@fmw42 : Hey, I tried that code already with the per-existing images with no luck. It produced an all gray image.

`	$img1 = new Imagick ('C:\Users\Mike\Documents\images\mask.png');;  //black circle mask png
	$img2 = new Imagick ('C:\Users\Mike\Documents\images\final.png');    //my read image png

	$img1->compositeImage( $img2, imagick::COMPOSITE_COPYOPACITY, 0, 0);
	$img1->writeImage('C:\Users\Mike\Documents\images\test.png');`
	
I throw everything in a function right now because I'm trying a bunch of stuff. If I wrap it in a function I don't have to comment out parts of the script and if an element works I can preserve it and run it again later easier until I figure out the issue.

I think you are onto something with the setImageAlphaChannel. I'll explore that next. I found in an old book from 2006 the mention of having to create the mask from the 'matte' channel. first the matte is extracted and then the alpha is created from that. I wonder if that isn't related to the link you posted. Thanks again for the leads though. Again, very much appreciated.

**Reply by fmw42** [reply]
You have it backwards. img1 should be the red image and img2 should be the mask. But see my other comment about putting the gray scale image into its own alpha channel first with setImageAlphaChannel and the shape option.

**Reply by Kramazeek** [reply]
@fmw42 - Got it! Read a bit of info in that link you sent and tried the legacy SetImageMatte function AFTER the draw call to turn off the transparency data. Without transparency data the COPY_OPACITY constant takes care of the rest. Thank you for all your help! You're the best, man.
`
//Create a black and white mask
$mask = new Imagick();
$mask->newImage(300, 300, "#ffffff");

//draw black circle
$draw = new ImagickDraw(); 
$draw->setStrokeWidth(1);
$draw->setStrokeOpacity(1);
$draw->setStrokeColor(new ImagickPixel("#000000"));
$draw->setFillColor(new ImagickPixel("#000000"));
$draw->circle(300/2, 300/2, 300/2, 300/2 + 100);
$mask->drawImage($draw);
$mask->setImageMatte(false);   //<-------This got it going!

//Create final image
$final = new Imagick();
$final->newImage(300, 300, "#FF0000");

$final->compositeImage($mask, imagick::COMPOSITE_COPYOPACITY, 0, 0);`
